google-cloud-firestore/src/main/java/com/google/cloud/firestore/telemetry/MetricsUtil.java

```diff
@@ -0,0 +1,254 @@
+import io.opentelemetry.api.metrics.Meter;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.metrics.SdkMeterProvider;
+import io.opentelemetry.sdk.metrics.SdkMeterProviderBuilder;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * A utility interface for trace collection. Classes that implement this interface may make their
+ * own design choices for how they approach trace collection. For instance, they may be no-op, or
+ * they may use a particular tracing framework such as OpenTelemetry.
+ */
+public class MetricsUtil {
+
+  private final FirestoreOptions firestoreOptions;
+  private OpenTelemetry openTelemetry;
+  private ApiTracerFactory otelApiTracerFactory;
+
+  private Meter meter;
+  private DoubleHistogram endToEndRequestLatency;
+  private DoubleHistogram firstResponseLatency;
+
+  /**
+   * Creates and returns an instance of the MetricsUtil class.
+   *
+   * @param firestoreOptions The FirestoreOptions object that is requesting an instance of
+   *     MetricsUtil.
+   * @return An instance of the MetricsUtil class.
+   */
+  public MetricsUtil(FirestoreOptions firestoreOptions) {
+    this.firestoreOptions = firestoreOptions;
+
+    // TODO(mila): this checking should be done inside FirestoreOptions builder, so that we can
+    // setApiTracerFactory
+    // to collect GAX layer metrics
+    boolean createEnabledInstance = firestoreOptions.getOpenTelemetryOptions().isMetricsEnabled();
+
+    // The environment variable can override options to enable/disable telemetry collection.
+    String enableMetricsEnvVar = System.getenv(ENABLE_METRICS_ENV_VAR);
+    if (enableMetricsEnvVar != null) {
+      if (enableMetricsEnvVar.equalsIgnoreCase("true")
+          || enableMetricsEnvVar.equalsIgnoreCase("on")) {
+        createEnabledInstance = true;
+      }
+      if (enableMetricsEnvVar.equalsIgnoreCase("false")
+          || enableMetricsEnvVar.equalsIgnoreCase("off")) {
+        createEnabledInstance = false;
+      }
+    }
+
+    // TODO(mila): re-assess if it should follow tracing's design: enabled/disabled MetricsUtil
+    if (createEnabledInstance) {
+      try {
+        createMetricsUtil(firestoreOptions);
+      } catch (IOException e) {
+        System.out.println(e);
+      }
+    }
+  }
+
+  private void createMetricsUtil(FirestoreOptions firestoreOptions) throws IOException {
+    // TODO(mila): re-locate this create otel instance logic to FirestoreOptions, as we might need
+    // 2 different OTELs based on customers config
+    this.openTelemetry = getDefaultOpenTelemetryInstance();
+
+    OpenTelemetryMetricsRecorder recorder =
+        new OpenTelemetryMetricsRecorder(openTelemetry, METER_NAME);
+
+    this.otelApiTracerFactory = new MetricsTracerFactory(recorder, createStaticAttributes());
+    registerMetrics();
+  }
+
+  private Map<String, String> createStaticAttributes() {
```

Comment:
nit: may be easier if this function just returns `Attributes`. (and get rid of `toOtelAttributes()`)

Reply:
`MetricsTracerFactory` accepts Map<String, String> to be added to GAX layer metrics, and same static attributes needs to be added to 4 firestore layer metrics as well.